src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// We only build the Source Generator in the netstandard target
+#if NETSTANDARD
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace CSharpSyntaxGenerator
+{
+    public abstract class CachingSourceGenerator : ISourceGenerator
+    {
+        private static Tuple<ImmutableArray<byte>, ImmutableArray<(string hintName, SourceText sourceText)>>? s_cachedResult;
+
+        protected abstract bool TryGetRelevantInput(in GeneratorExecutionContext context, out AdditionalText? input, out SourceText? inputText);
+
+        protected abstract bool TryGenerateSources(
+            AdditionalText input,
+            SourceText inputText,
+            out ImmutableArray<(string hintName, SourceText sourceText)> sources,
+            out ImmutableArray<Diagnostic> diagnostics,
+            CancellationToken cancellationToken);
+
+        public virtual void Initialize(GeneratorInitializationContext context)
```

Comment:
does this need to be virtual?

Reply:
Here, no. But could be helpful if someone copies this example.

Reply:
My personal preference is not to design for that case :)

Reply:
My hope is we have a better API before we need people copying stuff like this.

Reply:
(I agree though though no virtual here...)

Reply:
> we have a better API before we need people copying stuff like this

I would not expect the marginal benefits of an API providing only this feature to warrant requiring downstream consumers to upgrade to a newer version of the compiler. Even if the API existed, this is the pattern I would recommend.

Reply:
If a customer copy/pastes this, they can add virtual if appropriate.  I don't see why our coffee needs to have it. I'm also just not a fan of the: code it this way in case someone copies it.

Reply:
I have no objection to removing `virtual` and will do so next time I update this pull request.